core/domain/classifier_services.py

```diff
@@ -424,41 +375,240 @@ def delete_classifier_training_job(job_id):
+
+        for algorithm_id, job_id in python_utils.ZIP(
+                algorithm_ids_to_add, job_ids):
+            (
+                state_training_jobs_mapping.algorithm_ids_to_job_ids.
+                set_job_id_for_algorithm_id(algorithm_id, job_id))
+
+    if algorithm_ids_to_upgrade:
+        for algorithm_id in algorithm_ids_to_upgrade:
+            classifier_training_job = (
+                classifier_models.ClassifierTrainingJobModel.get_by_id(
+                    state_training_jobs_mapping.algorithm_ids_to_job_ids.
+                    get_job_id_for_algorithm_id(algorithm_id)))
+            classifier_training_job.algorithm_version = (
+                algorithm_id_to_algorithm_version[algorithm_id])
+            classifier_training_job.next_scheduled_check_time = (
+                datetime.datetime.utcnow())
+            classifier_training_job.status = feconf.TRAINING_JOB_STATUS_NEW
+            classifier_training_job.put()
+
+    if algorithm_ids_to_remove:
+        for algorithm_id in algorithm_ids_to_remove:
+            delete_classifier_training_job(
+                state_training_jobs_mapping.algorithm_ids_to_job_ids.
+                get_job_id_for_algorithm_id(algorithm_id))
+            (
+                state_training_jobs_mapping.algorithm_ids_to_job_ids.
+                remove_algorithm_id(algorithm_id))
+
+    state_training_jobs_mapping_model = (
+        classifier_models.StateTrainingJobsMappingModel.get_model(
+            exp_id, exp_version, state_name))
+    state_training_jobs_mapping.validate()
+    state_training_jobs_mapping_model.algorithm_ids_to_job_ids = (
+        state_training_jobs_mapping.algorithm_ids_to_job_ids.to_dict())
+    state_training_jobs_mapping_model.put()
+
+
+def get_classifier_training_job_maps(exp_id, exp_version, state_names):
+    """Gets the list of classifier training jobs mapped to algorithm id from
+    the exploration attributes.
 
     Args:
         exp_id: str. ID of the exploration.
         exp_version: int. The exploration version.
         state_names: list(str). The state names for which we retrieve the job.
 
     Returns:
-        list(ClassifierTrainingJob). Domain objects for the Classifier training
-        job model.
+        list(dict(str: ClassifierTrainingJob)). Domain objects for the
+        Classifier training job model for algorithm_ids.
     """
-    training_job_exploration_mapping_models = (
-        classifier_models.TrainingJobExplorationMappingModel.get_models(
+    state_training_jobs_mapping_models = (
+        classifier_models.StateTrainingJobsMappingModel.get_models(
             exp_id, exp_version, state_names))
-    job_ids = []
-    for mapping_model in training_job_exploration_mapping_models:
+    algorithm_ids_to_job_ids = []
+    for mapping_model in state_training_jobs_mapping_models:
         if mapping_model is None:
             continue
-        job_ids.append(mapping_model.job_id)
+        algorithm_ids_to_job_ids.append(
```

Comment:
I really think you can't have a list called algorithm_ids_to_job_ids containing elements which are each described basically as algorithm_ids_to_job_ids.

Perhaps algorithm_id_to_job_id_mapping and algorithm_id_to_job_id_mappings?

Reply:
Done